### ISSUE TITLE
Animation Blending

### ISSUE THREAD

**Issue by jasondalycan** [noname0310/babylon-mmd#8]
Hi,
Are there plans for animation blending?  For example, playing one animation, then blending into a second animation?

**Comment by noname0310** [comment]
MmdRuntime is designed to be compatible with Babylon.js's existing animation system.

It seems that your desired outcome involves manipulating a skinned mesh with MMD's runtime elements such as IK, and Physics and blending them into animations. In this case, by using `MmdRuntime.createMmdModel`, you can activate MMD's runtime and then utilize Babylon.js's animation system to animate the mesh. This approach would enable animation blending.

In conclusion, Babylon.js's animation system already includes blending, so utilizing it would solve the issue at hand.


If you're planning to blend multiple VMD animations, I can implement animation blending for you.

**Comment by noname0310** [comment]
If you provide me with detailed information about the problem situation you're currently trying to address and the data you intend to use, I'll be able to consider various solutions.

**Comment by jasondalycan** [comment]
Here's an example using Three.js MMDLoader where the MMD file transitions from idle -> walking -> running:

![chrome-capture-2023-7-7 (1)](https://github.com/noname0310/babylon-mmd/assets/14168215/ac39b3d3-1e9a-4e8e-b891-261a37db3c3e)

This uses animation blending.  Are you able to provide sample code showing how using `MmdRuntime.createMmdModel()` and Babylon's animation blending can do this?

**Comment by noname0310** [comment]
I have thought about this problem today, and I have come to the conclusion that it might be difficult to create a simple sample code for it.

Furthermore, I plan to support the following features in the near future to enable animation blending.

- Support for animation blending in `MmdRuntimeAnimation`.
- Load VMD files using Babylon.js `Animation` container.

From the information you've provided, it seems that ultimately you want to utilize MMD characters as player characters or NPC characters.

So, I will also provide automatic retargeting to allow humanoid animations to be used with MMD models.

**Comment by jasondalycan** [comment]
Thanks those will be very useful features.

**Comment by noname0310** [comment]
It is now possible to blend animation using Babylon.js animation runtime.

There is no document yet and the sample code is here:
https://github.com/noname0310/babylon-mmd/blob/main/src/Test/Scene/babylonAnimationBlendingTestScene.ts

Read the codes and ask any questions

**Comment by jasondalycan** [comment]
This is great!  Thanks!

**Comment by noname0310** [comment]
Since 0.27.0, it is possible to retarget any humanoid animation to an MMD skeleton.

I implemented this because I thought it might be in your potential requirements.

This will be documented quite a bit later, but if you're interested, ask in an issue and I'll show you how to use it in detail.

- Here is the code that uses the `AnimationRetargeter` to retarget the mixamo animation to the MMD skeleton and then serialize it to JSON to save it:
https://github.com/noname0310/babylon-mmd/blob/main/src/Test/Scene/humanoidAnimationTestScene.ts

- And here's the code to load the json serialized `AnimationGroup` into the mmd skeleton:
https://github.com/noname0310/babylon-mmd/blob/main/src/Test/Scene/babylonAnimationLoadTestScene.ts

- Here is what the retargeted animation looks like. The mixamo skeleton is on the right and the mmd skeleton is on the left:

  https://github.com/noname0310/babylon-mmd/assets/48761044/e8c075ae-8e3a-4830-90ca-88275903df17

And animation blending support in mmd custom animation runtime will be implemented quite later. The reasons are as follows:

- It's already possible with the animation runtime of babylon.js(and very easy to do since 0.26.0).

- Since most MMD animations were not designed with animation blending in mind, I figured that very few of the assets distributed in VMD format would be available for animation blending.

**Comment by noname0310** [comment]
The MMD animation runtime now also supports animation blending. It's frame-perfect animation blending, so you'll need a bit of code to apply it to locomotion.

Here is an example code if you want to try it out
https://github.com/noname0310/babylon-mmd/blob/main/src/Test/Scene/compositeAnimationTestScene.ts

If you have any questions, please open a new issue :)
